File: agent/packages/opencode/src/video/payload.ts
import * as path from "path"
import {
  collectImageRefs,
  collectVideoRefs,
  resolveAssetURL,
  stringValue,
} from "./assets"
import type { PromptDocument } from "./prompt"
import { parsePromptFile } from "./prompt"

export const DEFAULT_RATIO = "9:16"
export const DEFAULT_RESOLUTION = "720P"

export interface PayloadOptions {
  projectRoot: string
  allowNonOSS?: boolean
  allowTextOnly?: boolean
}

export interface VideoPayload {
  action: "generate"
  prompt: string
  duration: number
  ratio: string
  resolution: string
  sourceImageUrl?: string
  referenceImageUrls?: string[]
  sourceVideoUrls?: string[]
  continuationTailSeconds?: number
}

export async function buildPayload(promptPath: string, opts: PayloadOptions): Promise<VideoPayload> {
  const doc = await parsePromptFile(promptPath)
  return buildPayloadFromDocument(doc, opts)
}

export async function buildPayloadFromDocument(doc: PromptDocument, opts: PayloadOptions): Promise<VideoPayload> {
  const allowNonOSS = Boolean(opts.allowNonOSS)
  const promptDir = path.dirname(doc.path)
  const imageURLs: string[] = []
  for (const ref of collectImageRefs(doc.frontmatter)) {
    imageURLs.push(await resolveAssetURL(ref.value, ref.expected, opts.projectRoot, allowNonOSS, promptDir))
  }

  const videoURLs: string[] = []
  for (const ref of collectVideoRefs(doc.frontmatter)) {
    videoURLs.push(await resolveAssetURL(ref.value, ref.expected, opts.projectRoot, allowNonOSS, promptDir))
  }

  if (imageURLs.length === 0 && !opts.allowTextOnly) {
    throw new Error(
      "no image OSS URL found. The gateway workflow requires at least one source image URL; upload assets and fill assets.images first",
    )
  }

  const payload: VideoPayload = {
    action: "generate",
    prompt: doc.body,
    duration: parseDuration(doc.frontmatter.duration),
    ratio: stringValue(doc.frontmatter.ratio) ?? DEFAULT_RATIO,
    resolution: stringValue(doc.frontmatter.resolution) ?? DEFAULT_RESOLUTION,
  }

  if (imageURLs.length > 0) {
    payload.sourceImageUrl = imageURLs[0]
    if (imageURLs.length > 1) payload.referenceImageUrls = imageURLs.slice(1)
  }
  if (videoURLs.length > 0) payload.sourceVideoUrls = videoURLs
  if (doc.frontmatter.continuation_tail_seconds !== undefined) {
    payload.continuationTailSeconds = parseDuration(doc.frontmatter.continuation_tail_seconds)
  }
  return payload
}

export function parseDuration(value: unknown): number {
  if (value === null || value === undefined) return 12
  if (typeof value === "number" && Number.isFinite(value)) return Math.trunc(value)
  if (typeof value === "string") {
    const raw = value.trim().toLowerCase().replace(/s$/, "")
    const parsed = Number.parseInt(raw, 10)
    if (Number.isFinite(parsed)) return parsed
  }
  return 12
}

File: videoctl/internal/payload/payload.go
package payload

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/cdotlock/assets-produce/videoctl/internal/assets"
	"github.com/cdotlock/assets-produce/videoctl/internal/prompt"
)

const (
	DefaultRatio      = "9:16"
	DefaultResolution = "720P"
)

type Options struct {
	ProjectRoot   string
	AllowNonOSS   bool
	AllowTextOnly bool
}

func Build(promptPath string, opts Options) (map[string]any, error) {
	doc, err := prompt.ParseFile(promptPath)
	if err != nil {
		return nil, err
	}
	return BuildFromDocument(doc, opts)
}

func BuildFromDocument(doc *prompt.Document, opts Options) (map[string]any, error) {
	imageRefs := collectPayloadImageRefs(doc.Frontmatter)
	videoRefs := collectPayloadVideoRefs(doc.Frontmatter)

	imageURLs := make([]string, 0, len(imageRefs))
	for _, ref := range imageRefs {
		resolved, err := assets.ResolveAssetURL(ref, "image", opts.ProjectRoot, opts.AllowNonOSS)
		if err != nil {
			return nil, err
		}
		imageURLs = append(imageURLs, resolved)
	}

	videoURLs := make([]string, 0, len(videoRefs))
	for _, ref := range videoRefs {
		resolved, err := assets.ResolveAssetURL(ref, "video", opts.ProjectRoot, opts.AllowNonOSS)
		if err != nil {
			return nil, err
		}
		videoURLs = append(videoURLs, resolved)
	}

	if previousFrame, ok := assets.StringValue(doc.Frontmatter["previous_frame_url"]); ok {
		resolved, err := assets.ResolveAssetURL(previousFrame, "previous_frame_url", opts.ProjectRoot, opts.AllowNonOSS)
		if err != nil {
			return nil, err
		}
		imageURLs = appendUnique(imageURLs, resolved)
	}

	if previousVideo, ok := assets.StringValue(doc.Frontmatter["previous_video_url"]); ok {
		resolved, err := assets.ResolveAssetURL(previousVideo, "previous_video_url", opts.ProjectRoot, opts.AllowNonOSS)
		if err != nil {
			return nil, err
		}
		videoURLs = appendUnique(videoURLs, resolved)
	}

	if len(imageURLs) == 0 && !opts.AllowTextOnly {
		return nil, fmt.Errorf("no image OSS URL found. The gateway workflow requires at least one source image URL; upload assets and fill assets.images first")
	}

	duration, err := parseDuration(doc.Frontmatter["duration"])
	if err != nil {
		return nil, err
	}

	out := map[string]any{
		"action":     "generate",
		"prompt":     doc.Body,
		"duration":   duration,
		"ratio":      stringDefault(doc.Frontmatter["ratio"], DefaultRatio),
		"resolution": stringDefault(doc.Frontmatter["resolution"], DefaultResolution),
	}

	if len(imageURLs) > 0 {
		out["sourceImageUrl"] = imageURLs[0]
		if len(imageURLs) > 1 {
			out["referenceImageUrls"] = imageURLs[1:]
		}
	}

	if len(videoURLs) > 0 {
		out["sourceVideoUrls"] = videoURLs
	}

	if value, ok := doc.Frontmatter["continuation_tail_seconds"]; ok && value != nil {
		tailSeconds, err := parseDuration(value)
		if err != nil {
			return nil, err
		}
		out["continuationTailSeconds"] = tailSeconds
	}

	return out, nil
}

func collectPayloadImageRefs(frontmatter map[string]any) []string {
	var refs []string
	for _, field := range []string{"first_frame", "last_frame"} {
		if value, ok := assets.StringValue(frontmatter[field]); ok {
			refs = append(refs, value)
		}
	}

	assetMap := assets.MapValue(frontmatter["assets"])
	refs = append(refs, assets.StringSlice(assetMap["images"])...)
	return refs
}

func collectPayloadVideoRefs(frontmatter map[string]any) []string {
	assetMap := assets.MapValue(frontmatter["assets"])
	return assets.StringSlice(assetMap["videos"])
}

func parseDuration(value any) (int, error) {
	if value == nil {
		return 12, nil
	}

	switch typed := value.(type) {
	case int:
		return typed, nil
	case int64:
		return int(typed), nil
	case string:
		raw := strings.TrimSpace(strings.ToLower(typed))
		raw = strings.TrimSuffix(raw, "s")
		return strconv.Atoi(raw)
	default:
		return 12, nil
	}
}

func stringDefault(value any, fallback string) string {
	text, ok := assets.StringValue(value)
	if !ok {
		return fallback
	}
	return text
}

func appendUnique(values []string, value string) []string {
	for _, existing := range values {
		if existing == value {
			return values
		}
	}
	return append(values, value)
}
